File: TrainBlink/android/app/src/main/java/com/trainblink/app/models/Station.kt
package com.trainblink.app.models

import com.google.android.gms.maps.model.LatLng

/**
 * Station Model
 * Represents a train station (TRA or THSR)
 * Android equivalent of iOS Station.swift
 */
data class Station(
    val id: String,
    val name: String,
    val type: StationType,
    val location: LatLng,
    val geofenceRadiusMeters: Double = 500.0,
    val address: String? = null,
    val isActive: Boolean = true
) {
    companion object {
        // Sample TRA stations
        val taipeiMain = Station(
            id = "TRA-1000",
            name = "Taipei Main Station",
            type = StationType.TRA,
            location = LatLng(25.0478, 121.5170),
            address = "No. 3, Section 1, Zhongxiao West Road, Zhongzheng District, Taipei"
        )

        val banqiao = Station(
            id = "TRA-1020",
            name = "Banqiao Station",
            type = StationType.TRA,
            location = LatLng(25.0140, 121.4627)
        )

        val songshan = Station(
            id = "TRA-1001",
            name = "Songshan Station",
            type = StationType.TRA,
            location = LatLng(25.0494, 121.5778)
        )

        // Sample THSR stations
        val taipeiTHSR = Station(
            id = "THSR-01",
            name = "Taipei THSR Station",
            type = StationType.THSR,
            location = LatLng(25.0478, 121.5170)
        )

        val banqiaoTHSR = Station(
            id = "THSR-02",
            name = "Banqiao THSR Station",
            type = StationType.THSR,
            location = LatLng(25.0140, 121.4627)
        )

        val taichungTHSR = Station(
            id = "THSR-06",
            name = "Taichung THSR Station",
            type = StationType.THSR,
            location = LatLng(24.1129, 120.6166)
        )

        // All sample stations
        val samples = listOf(
            taipeiMain, banqiao, songshan,
            taipeiTHSR, banqiaoTHSR, taichungTHSR
        )
    }

    /**
     * Distance from given location in meters
     */
    fun distanceFrom(lat: Double, lng: Double): Double {
        val results = FloatArray(1)
        android.location.Location.distanceBetween(
            lat, lng,
            location.latitude, location.longitude,
            results
        )
        return results[0].toDouble()
    }

    /**
     * Whether a location is within geofence radius
     */
    fun isWithinGeofence(lat: Double, lng: Double): Boolean {
        return distanceFrom(lat, lng) <= geofenceRadiusMeters
    }
}

/**
 * Station Type
 */
enum class StationType(val displayName: String) {
    TRA("TRA"),
    THSR("THSR");

    companion object {
        fun fromString(value: String): StationType {
            return values().find { it.name.equals(value, ignoreCase = true) } ?: TRA
        }
    }
}

File: TrainBlink/ios/TrainBlink/Models/Station.swift
//
//  Station.swift
//  TrainBlink
//
//  Feature 1: Geofencing System
//  Station data model
//
//  Created by TrainBlink Team
//  Copyright © 2025 TrainBlink. All rights reserved.
//

import Foundation
import CoreLocation

/// Station type classification
enum StationType: String, Codable {
    case tra = "TRA"        // 台鐵 Taiwan Railways
    case thsr = "THSR"      // 高鐵 Taiwan High Speed Rail
    case mrt = "MRT"        // 捷運 Metro (future)
}

/// Station model representing a train station
struct Station: Codable, Identifiable, Hashable {

    // MARK: - Properties

    let id: String              // Unique station ID (e.g., "1000" for Taipei)
    let name: String            // Traditional Chinese name (e.g., "台北車站")
    let nameEn: String          // English name (e.g., "Taipei")
    let latitude: Double        // GPS latitude
    let longitude: Double       // GPS longitude
    let type: StationType       // Station type (TRA/THSR/MRT)
    let radius: Double          // Geofence radius in meters (default: 500m)
    let lines: [String]?        // Railway lines (e.g., ["西部幹線", "東部幹線"])

    // MARK: - Coding Keys

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case nameEn = "name_en"
        case latitude = "lat"
        case longitude = "lng"
        case type
        case radius
        case lines
    }

    // MARK: - Computed Properties

    /// CLLocationCoordinate2D for MapKit/CoreLocation
    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }

    /// CLLocation for distance calculations
    var location: CLLocation {
        CLLocation(latitude: latitude, longitude: longitude)
    }

    /// Display name with type badge
    var displayName: String {
        "\(name) (\(type.rawValue))"
    }

    // MARK: - Methods

    /// Calculate distance from given location
    func distance(from location: CLLocation) -> CLLocationDistance {
        return self.location.distance(from: location)
    }

    /// Check if location is within station radius
    func contains(_ location: CLLocation) -> Bool {
        return distance(from: location) <= radius
    }

    /// Create a geofence region for this station
    func makeGeofenceRegion() -> CLCircularRegion {
        let region = CLCircularRegion(
            center: coordinate,
            radius: radius,
            identifier: id
        )
        region.notifyOnEntry = true
        region.notifyOnExit = true
        return region
    }
}

// MARK: - Sample Data

extension Station {

    /// Sample stations for testing and previews
    static let samples: [Station] = [
        Station(
            id: "1000",
            name: "台北車站",
            nameEn: "Taipei",
            latitude: 25.0478,
            longitude: 121.5170,
            type: .tra,
            radius: 500,
            lines: ["西部幹線", "東部幹線"]
        ),
        Station(
            id: "THSR-1",
            name: "台北站",
            nameEn: "Taipei",
            latitude: 25.0478,
            longitude: 121.5170,
            type: .thsr,
            radius: 500,
            lines: ["高鐵"]
        ),
        Station(
            id: "1100",
            name: "台中車站",
            nameEn: "Taichung",
            latitude: 24.1369,
            longitude: 120.6850,
            type: .tra,
            radius: 500,
            lines: ["西部幹線"]
        ),
        Station(
            id: "THSR-3",
            name: "台中站",
            nameEn: "Taichung",
            latitude: 24.1123,
            longitude: 120.6173,
            type: .thsr,
            radius: 500,
            lines: ["高鐵"]
        ),
        Station(
            id: "1200",
            name: "高雄車站",
            nameEn: "Kaohsiung",
            latitude: 22.6391,
            longitude: 120.3018,
            type: .tra,
            radius: 500,
            lines: ["西部幹線", "南迴線"]
        )
    ]
}
